### DISCUSSION TITLE
Obtaining shadow price (marginal cost) of different commodities

### DISCUSSION THREAD

**Discussion by zsalimian** [iiasa/message_ix#427] [MESSAGE modeling] [answered]
Dear @khaeru 
Hi,
I need to edit objective function of message-ix so that it could maximize social welfare instead of minimizing the total production costs.
I need the marginal cost of each commodity. For instance, in simple westeros example I would like to see the amount of shadow price or marginal cost of light commodity.So, I have the following questions:
1. Is 𝑃𝑅𝐼𝐶𝐸_𝐶𝑂𝑀𝑀𝑂𝐷𝐼𝑇Y represent the shadow price of a commodity, such as light,in the dual solution?
2. Probably, it is the shadow price of the commodity balance constraint. But I confuesed bacause I saw we had two constarints for commodity balance, including commodity_Balance_GT and Commodity_balance_LT.
3.I run westeros in python. How can I call shadow price of light in python version of westeros and print its amount?

Regards

**Comment by khaeru** [comment]
These questions can all be answered by reading the documentation, tutorials, and GAMS code.

Please:
- For each of the three issues #416, #419, and #423 that you previously opened: if you have resolved them, then mark them as closed.
- Do not tag specific message_ix contributors & developers when filing issues. Issues are handled jointly by the development team.

**Comment by behnam-zakeri** [comment, marked as answer]
@zsalimian thanks for your questions. I answer with the same order:
1. Yes, that's true.
2. In the GAMS formulation, in the file `message_ix\model\MESSAGE\message_solve.gms`, you can find [the calculation of `PRICE_COMMODITY`](https://github.com/iiasa/message_ix/blob/538e1ef24f06790ca0c33516b79da970ae40a77b/message_ix/model/MESSAGE/model_solve.gms#L52). It is based on both the marginals of commodity balance for "greater than" (GT) and "lower than" (LT) equations, divided by discount factor. If the marginal is zero in one of the above-mentioned equations, the marginal of the other one defines the price of commodity.
3. You should be able to look into the content of an output variable by calling `scenario.var()`, after solving the scenario successfully. For example, for this case, you can do: `df = scenario.var('PRICE_COMMODITY')`. Then, you can see the results in python through the resulting dataframe (df).
I hope it helps.

**Reply by meng25meng** [reply]
> 3\. You should be able to look into the content of an output variable by calling `scenario.var()`, after solving the scenario successfully. For example, for this case, you can do: `df = scenario.var('PRICE_COMMODITY')`. Then, you can see the results in python through the resulting dataframe (df).
>     I hope it helps.

HELLLO，i also encountered a similar problem the error is as follows:  commmodity balance g is 0，which parament should be modified? i did not add a bound-activity lo or up，
Infeasibility row 'COMMODITY_BALANCE_GT(HLJ.electricity.final.2025.year)':  0 >= 14.2191.
Presolve time = 0.02 sec. (10.31 ticks)
LP status(3): infeasible
Cplex Time: 0.02sec (det. 10.39 ticks)

**Reply by glatterf42** [reply]
Hello @meng25meng, this comment seems related to https://github.com/iiasa/message_ix/discussions/846#discussioncomment-9826625, is this correct? 
Please help us keep an overview of this repository :pray: 
This means: for every error you encounter or question you have, open a new issue or discussion. Once this issue is resolved or discussion answered, please mark them as such. 
Doing this will help us provide faster and more targeted help, so thanks in advance!

**Comment by zsalimian** [comment]
Thank you for your kind reply.
I saw message_ix\model\MESSAGE\message_solve.gms and it was helpful alot.
As you mentioned above, I used df = scenario.var('PRICE_COMMODITY') after scenario.solve() and then call print.
It should me the following output
       node    commodity      level  year  time          lvl            mrg
0  Westeros  electricity  secondary   700  year  405.643395  0.0
1  Westeros  electricity  secondary   710  year   90.980599  0.0
2  Westeros  electricity  secondary   720  year   90.980599  0.0
3  Westeros  electricity      final        700  year  506.270438  0.0
4  Westeros  electricity      final        710  year  156.645110  0.0
5  Westeros  electricity      final        720  year  156.645110  0.0
6  Westeros        light     useful        700  year  511.270438  0.0
7  Westeros        light     useful        710  year  161.645110  0.0
8  Westeros        light     useful        720  year  161.645110  0.0

I don't need the marginal cost of balance equations seprately. I think I can use the lvl column for the amount of shadow price. So, I think the lvl column shows the marginal price. Isn't it?
But how about mrg. I saw in the docs that ‘mrg’ and ‘lvl’ are valid only for ix_type='equ', and ‘level’ otherwise.It said it is related to level and equations. But I didn't underestand why it is equal to zero.
Regards

**Comment by behnam-zakeri** [comment]
Please be specific with your quesitons. Are you asking about `lvl` and `mrg` of the equations of `COMMODITY_BALANCE` or variable `PRICE_COMMODITY`? As mentioned before, `PRICE_COMMODITY` is a variable calculated by using marginals (`mrg`) of the equations of commodity balance. So, in fact, this calculation is done outside the optimization problem, in a potprocessing step. As such, the marginals for `PRICE_COMMODITY` are not needed (`mrg=0`), and they are meaningless. Because the value of `PRICE_COMMODITY` doesn't have any impact on the model objective function. I hope this helps.

**Comment by zsalimian** [comment]
Dear Dr.,
Many many thanks for your reply.
Please let me clarify my question. 
I mentioned before that I need to edit objective function of message-ix so that it could maximize social welfare instead of minimizing the total production costs. I told that I need the marginal price of the energy commodities for editing the objective function. 
**_### I understood from your answer I can use the PRICE_COMMODITY as marginal cost because it is average of shadow prices of balance equations._** 
Then you guided me to use df = scenario.var('PRICE_COMMODITY') syntax in python code. I executed the code, and above output was produced. Then I confused whether the reults in the lvl column shows the PRICE_COMMODITY or not?For instance, is the PRICE_COMMODITY of light in the last year equal to 161.645110 or not?(If possible please see the above output)
### **_If your answer is 'yes',then 161.645110 is the PRICE_COMMODITY of light in the last year. So it represents the marginal cost of light in the last year and it must be the average of shadow prices of balance equations._**
In other words, as I mentioned before the marginal cost of commodity is sufficient for editing objective functions in GAMS code, but I confused which of them ('lvl' or 'mrg' columns) in the **_above output_** represent marginal cost, which is the average of shadow price of balance equations.
Many thanks in advance for your kind answer

**Comment by behnam-zakeri** [comment]
Thanks for the clarification. The answer to your question in general is "yes". I correct your sentence as "`PRICE_COMMODITY` levels, given under `lvl`, represents **the marginal cost of meeting demand for commodity** `light`".

**Reply by zsalimian** [reply]
Dear Dr. Zakeri,
For more accuracy I have to ask this question. In the MESSAGE-ix documentation it is mentioned that the "PRICE_COMMODITY" is undiscounted. Moreover, I saw in the documentation that sum of shadow prices of both balance equations is divided to discount rate for undiscounting. Why does this division happen? I contrast this by TIMES model. In TIMES it is mentioned that although the objective function is discounted, the shadow prices of constraints are already undiscounted and we don't need to divide them by discount rates. I think it should be the same in MESSAGE. May be I make a mistake. Would you please guide me? 

Moreover, I would like to add the amount of interest rate to the westeros model to calculate the discount factor as follow:
df=1/(1+dr)^i
I saw the following syntax but I don't know how to introduce the amount of interestrate in the model.
scenario.add_par("interestrate", model_horizon, value=0.05, unit='-')

With Best Regards
Zohreh
